src/audit/stale_action_refs.rs

```diff
@@ -0,0 +1,92 @@
+//! Detects actions pinned by commit hash, which doesn't point to a Git tag.
+
+use anyhow::{Context, Result, anyhow};
+use github_actions_models::common::{RepositoryUses, Uses};
+
+use super::{Audit, AuditLoadError, audit_meta};
+use crate::{
+    Persona,
+    finding::{Confidence, Finding, Severity},
+    github_api,
+    models::{CompositeStep, Step, StepCommon, uses::RepositoryUsesExt as _},
+    state::AuditState,
+};
+
+pub(crate) struct StaleActionRefs {
+    client: github_api::Client,
+}
+
+audit_meta!(
+    StaleActionRefs,
+    "stale-action-refs",
+    "commit hash does not point to a Git tag"
+);
+
+impl StaleActionRefs {
+    fn is_stale_action_ref(&self, uses: &RepositoryUses) -> Result<bool> {
+        let tag = match &uses.commit_ref() {
+            Some(commit_ref) => self
+                .client
+                .longest_tag_for_commit(&uses.owner, &uses.repo, commit_ref)
+                .with_context(|| {
+                    format!(
+                        "couldn't retrieve tag for {owner}/{repo}@{commit_ref}",
+                        owner = uses.owner,
+                        repo = uses.repo
+                    )
+                })?,
```

Comment:
Nitpick: this is a great context message to add, but it's so good that it makes me realize that it should be in `longest_tag_for_commit` itself 🙂 -- could you move it there?

Reply:
Thanks for the compliment, but I just copied the message from `src/audit/known_vulnerable_actions.rs`, and _you_ were the one who added it there originally 😅

The `with_context` call there is now probably redundant; have created #729 to remove it.

Reply:
🤦